wrappers/python/nntile/layer/llama_attention.py

```diff
@@ -1517,3 +1517,156 @@ def to_torch_with_grads(self) -> LlamaAttention_torch:
+        a_flops = 2 * np.prod(k_rep_shape[:2]) * np.prod(q_rope_shape[1:])
+        total_forward_flops += a_flops
+        # Apply value tensor
+        # B = einsum('jklbi,kmlbi->jmlbi', V_rep, A)
+        # batched gemm
+        # (head_size, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        # by (n_seq, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        # into (head_size, n_seq, batch=(n_batch, kv_group_size, n_head_kv))
+        v_rep_shape = self.v_rep.value.shape
+        a_shape = self.a.value.shape
+        b_flops = 2 * np.prod(a_shape) * v_rep_shape[0]
+        total_forward_flops += b_flops
+        # Gemm for accumulate result for all the heads
+        # Y = einsum('jklm,klmni->jni', W, B_transposed)
+        # gemm (n_emb, kv_group_size, n_head_kv, head_size) by
+        # (kv_group_size, n_head_kv, head_size, n_seq, n_batch)
+        # into (n_emb, n_seq, n_batch)
+        w_shape = self.w.value.shape
+        bt_shape = self.b_transposed.value.shape
+        y_flops = 2 * np.prod(w_shape) * np.prod(bt_shape[3:])
+        total_forward_flops += y_flops
+        return total_forward_flops
+
+    def get_layer_backward_flops(self):
+        total_backward_flops = 0
+        if self.w.grad_required:
+            # dW += einsum('jni,klmni->jklm', dY, B_transposed)
+            y_grad_shape = self.y.grad.shape
+            bt_shape = self.b_transposed.value.shape
+            w_grad_flops = 2 * np.prod(bt_shape) * y_grad_shape[:-2]
+            total_backward_flops += w_grad_flops
+        if self.b_transposed.grad_required:
+            # dB_transposed = einsum('jklm,jni->klmni', W, dY)
+            w_shape = self.w.value.shape
+            y_grad_shape = self.y.grad.shape
+            bt_grad_flops = 2 * np.prod(w_shape) * np.prod(y_grad_shape[1:])
+            total_backward_flops += bt_grad_flops
+        if self.a.grad_required:
+            # dA = einsum('jklbi,jmlbi->kmlbi', V_rep, dB)
+            # ndim = 1
+            # batch_ndim = 3
+            b_grad_shape = self.b.grad.shape
+            v_rep_shape = self.v_rep.value.shape
+            a_grad_flops = 2 * np.prod(v_rep_shape) * b_grad_shape[1]
+            total_backward_flops += a_grad_flops
+        if self.v_rep.grad_required:
+            # dV_rep = einsum('jmlbi,kmlbi->jklbi', dB, A)
+            # ndim = 1
+            # batch_ndim = 3
+            a_shape = self.a.value.shape
+            b_grad_shape = self.b.grad.shape
+            v_rep_grad_flops = 2 * np.prod(a_shape) * b_grad_shape[0]
+            total_backward_flops += v_rep_grad_flops
+        if self.k_rep.grad_required:
+            # dK_rep = 1.0/sqrt(head_size)
+            #          * einsum('jmlbi,kmlbi->jklbi', Q_rope, dA)
+            # ndim = 1
+            # batch_ndim = 3
+            q_rope_shape = self.q_rope.value.shape
+            a_grad_shape = self.a.grad.shape
+            k_rep_grad_flops = 2 * np.prod(a_grad_shape) * q_rope_shape[1]
```

Comment:
judging by `einsum` there is mistake: `a_grad_shape` does not contain index `j`, which is `q_rope_shape[0]` instead of `q_rope_shape[1]`.

Reply:
`a_grad_shape` is `kmlbi`, not `jklbi`.